### PR TITLE
Attempt reconstruction from all compact block announcements

### PR DESCRIPTION
**Pull request by sdaftuar**
Previously, once a block was in flight, we would ignore compact block announcements from our high-bandwidth peers, even though the block might reconstruct with no additional bandwidth required.

This allows us to optimistically try to reconstruct a compact block from our peers, and if it succeeds with no round-trip, then process the block.

### CODE REVIEW

**Comment by gmaxwell** [comment]
Concept ACK, good step forward.

**Comment by rebroad** [comment]
#9325 already does this, and more.

**Comment by sdaftuar** [comment]
@rebroad Responded in #9325 about some of the issues with that patch.

**Comment by gmaxwell** [comment]
Testing.

**Comment by instagibbs** [comment]
utACK 7017298eb2c72e262b865cd417d6844bcdd14a3f

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1889,6 +1911,29 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
+
+        if (fBlockReconstructed) {
+            // If we got here, we were able to optimistically reconstruct a
+            // block that is in flight from some other peer.
+            {
+                LOCK(cs_main);
+                mapBlockSource.emplace(pblock->GetHash(), std::make_pair(pfrom->GetId(), false));
+            }
+            bool fNewBlock = false;
+            ProcessNewBlock(chainparams, pblock, true, &fNewBlock);
```
Comment (TheBlueMatt) [inline src/net_processing.cpp]: The documentation is just flat out wrong, it seems. fForceProcessing means "make sure to store this to disk, we decided we wanted to request this block for some reason, so probably want it no matter what validation.cpp thinks". Indeed, the use here is slightly different from elsewhere, but given that we already check for blocks which do not mesh with our idea of the best chain further up, and that we will only get here if the block was within two blocks of our tip, I'm pretty confident this is the right use.

**Comment by TheBlueMatt** [comment]
utACK 813ede91e12dd777faee60f6286a9ed5bb72761a

**Comment by gmaxwell** [comment]
ACK. Appears to be highly effective on mainnet too.

**Comment by btcdrak** [comment]
utACK 7017298eb2c72e262b865cd417d6844bcdd14a3f

**Comment by laanwj** [comment]
This was backported in #9357

**Comment by morcos** [comment]
posthumous utACK

**Pull request merged**